src/mjolnir/graphfilter.cc

```diff
@@ -27,10 +29,157 @@ uint32_t n_original_nodes = 0;
 uint32_t n_filtered_edges = 0;
 uint32_t n_filtered_nodes = 0;
 uint32_t can_aggregate = 0;
+uint32_t aggregated = 0;
 
 // Group wheelchair and pedestrian access together
 constexpr uint32_t kAllPedestrianAccess = (kPedestrianAccess | kWheelchairAccess);
 
+struct EdgeId {
+  uint64_t way_id;
+  GraphId graph_id;
+};
+
+bool CanAggregate(const DirectedEdge* de) {
+  if (de->start_restriction() || de->part_of_complex_restriction() || de->end_restriction() ||
+      de->restrictions() || de->traffic_signal() || de->access_restriction()) {
+    return false;
+  }
+  return true;
+}
+
+bool ExpandFromNode(GraphReader& reader,
+                    std::list<PointLL>& shape,
+                    GraphId& en,
+                    const GraphId& from_node,
+                    std::unordered_set<std::string>& isos,
+                    bool forward,
+                    GraphId& last_node,
+                    std::unordered_set<GraphId>& visited_nodes,
+                    uint64_t& way_id,
+                    const graph_tile_ptr& prev_tile,
+                    GraphId prev_node,
+                    GraphId current_node);
+
+bool ExpandFromNodeInner(GraphReader& reader,
+                         std::list<PointLL>& shape,
+                         GraphId& en,
+                         const GraphId& from_node,
+                         std::unordered_set<std::string>& isos,
+                         bool forward,
+                         GraphId& last_node,
+                         std::unordered_set<GraphId>& visited_nodes,
+                         uint64_t& way_id,
+                         const graph_tile_ptr& prev_tile,
+                         GraphId prev_node,
+                         GraphId current_node,
+                         const NodeInfo* node_info) {
+
+  for (size_t j = 0; j < node_info->edge_count(); ++j) {
+    GraphId edge_id(prev_tile->id().tileid(), prev_tile->id().level(), node_info->edge_index() + j);
+    const DirectedEdge* de = prev_tile->directededge(edge_id);
+    auto edge_info = prev_tile->edgeinfo(de);
```

Comment:
can we live with a reference here instead of a copy or is that hidden in the `auto`? if it's implicit, let's add the `&` so it's clearer